services/bls_aggregation/blsagg.go

```diff
@@ -0,0 +1,364 @@
+	go a.singleTaskAggregatorGoroutineFunc(taskIndex, taskCreatedBlock, quorumNumbers, quorumThresholdPercentages, timeToExpiry, taskC)
+	return nil
+}
+
+func (a *BlsAggregatorService) ProcessNewSignature(
+	ctx context.Context,
+	taskIndex types.TaskIndex,
+	taskResponseDigest types.TaskResponseDigest,
+	blsSignature *bls.Signature,
+	operatorId bls.OperatorId,
+) error {
+	a.taskChansMutex.Lock()
+	taskC, taskInitialized := a.taskChans[taskIndex]
+	a.taskChansMutex.Unlock()
+	if !taskInitialized {
+		return TaskNotFoundErrorF(taskIndex)
+	}
+	signatureVerificationErrorC := make(chan error)
+	// send the task to the goroutine processing this task
+	// and return the error (if any) returned by the signature verification routine
+	select {
+	case taskC <- types.SignedTaskResponseDigest{
+		TaskResponseDigest:          taskResponseDigest,
+		BlsSignature:                blsSignature,
+		OperatorId:                  operatorId,
+		SignatureVerificationErrorC: signatureVerificationErrorC,
+	}:
+		// note that we need to wait synchronously here for this response because we want to
+		// send back an informative error message to the operator who sent his signature to the aggregator
+		return <-signatureVerificationErrorC
+	case <-ctx.Done():
+		return ctx.Err()
+	}
+}
+
+func (a *BlsAggregatorService) singleTaskAggregatorGoroutineFunc(
+	taskIndex types.TaskIndex,
+	taskCreatedBlock uint32,
+	quorumNumbers []types.QuorumNum,
+	quorumThresholdPercentages []types.QuorumThresholdPercentage,
+	timeToExpiry time.Duration,
+	taskRespC <-chan types.SignedTaskResponseDigest,
+) {
+	defer a.closeTaskGoroutine(taskIndex)
+
+	quorumThresholdPercentagesMap := make(map[types.QuorumNum]types.QuorumThresholdPercentage)
+	for i, quorumNumber := range quorumNumbers {
+		quorumThresholdPercentagesMap[quorumNumber] = quorumThresholdPercentages[i]
+	}
+	operatorsAvsStateDict, err := a.avsRegistryService.GetOperatorsAvsStateAtBlock(context.Background(), quorumNumbers, taskCreatedBlock)
+	if err != nil {
+		// TODO: how should we handle such an error?
+		a.logger.Fatal("Aggregator failed to get operators state from avs registry", "err", err)
+	}
+	quorumsAvsStakeDict, err := a.avsRegistryService.GetQuorumsAvsStateAtBlock(context.Background(), quorumNumbers, taskCreatedBlock)
+	if err != nil {
+		a.logger.Fatal("Aggregator failed to get quorums state from avs registry", "err", err)
+	}
+	totalStakePerQuorum := make(map[types.QuorumNum]*big.Int)
+	for quorumNum, quorumAvsState := range quorumsAvsStakeDict {
+		totalStakePerQuorum[quorumNum] = quorumAvsState.TotalStake
+	}
+	quorumApksG1 := []*bls.G1Point{}
+	for _, quorumNumber := range quorumNumbers {
+		quorumApksG1 = append(quorumApksG1, quorumsAvsStakeDict[quorumNumber].AggPubkeyG1)
+	}
+
+	// TODO(samlaf): instead of taking a TTE, we should take a block as input
+	// and monitor the chain and only close the task goroutine when that block is reached
+	taskExpiredTimer := time.NewTimer(timeToExpiry)
```

Comment:
reading your comment, how does the timer would read what is the current block number? Seems like there would be need for a timer routine which reads current block number which it then sends via chan to each of the active singleTaskAggregatorGoroutineFunc. wdyt?

Reply:
You'd just set a ticker that ticks every 12s (roughly once per block), queries the chain for blockNum, check if its reached the TTEBlock, and if so close the goroutine.